core/sector_storage/impl/fetch_handler.cpp

```diff
@@ -0,0 +1,276 @@
+    }
+
+    auto response = std::make_shared<http::response<http::dynamic_body>>();
+    setupResponse(request, response);
+    response->set(http::field::content_type, "application/json");
+    boost::beast::ostream(response->body())
+        .write(fc::common::span::bytestr(maybe_json.value().data()),
+               maybe_json.value().size());
+    doWrite(socket, response);
+    return api::WrapperResponse{
+        .response = std::move(response),
+    };
+  }
+
+  api::WrapperResponse remoteGetSector(
+      tcp::socket &socket,
+      const http::request<http::dynamic_body> &request,
+      const std::shared_ptr<stores::LocalStore> &local_store,
+      const common::Logger &logger,
+      const std::string &type,
+      const std::string &sector) {
+    auto maybe_sector = fc::primitives::sector_file::parseSectorName(sector);
+
+    if (maybe_sector.has_error()) {
+      logger->error("Error remote get sector: {}",
+                    maybe_sector.error().message());
+      return makeErrorResponse(
+          socket, request, http::status::internal_server_error);
+    }
+
+    auto maybe_type = fc::primitives::sector_file::fromString(type);
+    if (maybe_type.has_error()) {
+      logger->error("Error remote get sector: {}",
+                    maybe_type.error().message());
+      return makeErrorResponse(
+          socket, request, http::status::internal_server_error);
+    }
+
+    // Proof type is 0 because we don't allocate anything
+    auto maybe_paths = local_store->acquireSector(
+        maybe_sector.value(),
+        static_cast<RegisteredProof>(0),
+        maybe_type.value(),
+        fc::primitives::sector_file::SectorFileType::FTNone,
+
+        fc::sector_storage::stores::PathType::kStorage,
+        fc::sector_storage::stores::AcquireMode::kMove);
+    if (maybe_paths.has_error()) {
+      logger->error("Error remote get sector: {}",
+                    maybe_paths.error().message());
+      return makeErrorResponse(
+          socket, request, http::status::internal_server_error);
+    }
+
+    auto maybe_path =
+        maybe_paths.value().paths.getPathByType(maybe_type.value());
+    if (maybe_path.has_error()) {
+      logger->error("Error remote get sector: {}",
+                    maybe_path.error().message());
+      return makeErrorResponse(
+          socket, request, http::status::internal_server_error);
+    }
+    if (maybe_path.value().empty()) {
+      logger->error("Error remote get sector: acquired path was empty");
+      return makeErrorResponse(
+          socket, request, http::status::internal_server_error);
+    }
+
+    std::string response_file = maybe_path.value();
+
+    auto response = std::make_shared<http::response<http::file_body>>();
+    setupResponse(request, response);
+    if (boost::filesystem::is_directory(maybe_path.value())) {
+      auto temp_file = fs::temp_directory_path() / fs::unique_path();
+      auto _ = gsl::finally([&temp_file, &logger]() {
```

Comment:
Can `gsl::finally` remove tar before fully sending?
Or will file content survive because it has opened file descriptors?

Reply:
Fixed